### ISSUE TITLE
UPnP not discovered

### ISSUE THREAD

**Issue by jpritikin** [ethereum/go-ethereum#890]
I built v0.9.18 from source on linux amd64.

I0509 13:47:01.427490   19131 nat.go:94] network port 30303 could not be mapped: no UPnP or NAT-PMP router discovered

However, I'm pretty sure my router has UPnP enabled, http://i.imgur.com/mf3xcD4.png

**Comment by obscuren** [comment]
Routers don't give descriptive error messages unfortunately :-( There can be a whole range of things that have happened when it tried to map the port. UPnP uses SOAP over HTTP and my router for example occasionally decides to give 500 errors which also gives this exact error. Unfortunately this is a very common problem and there isn't a lot we can do :-(

**Comment by jpritikin** [comment]
eth from cpp-ethereum works fine,

````
  ℹ  06:00:38|p2p  UPnP device: http://192.168.2.1:1780/InternetGatewayDevice.xml [st: urn:schemas-upnp-org:device:InternetGatewayDevice:1 ]
⧎ ℹ  06:00:38|p2p  Punched through NAT and mapped local port 30303 onto external port 30303 .
````

To help you track down the problem, I made wireshark recordings. Here is geth,

https://drive.google.com/file/d/0BwpGqX45kuYIU29OLTJTWWhlOWs/view?usp=sharing

and here is eth

https://drive.google.com/file/d/0BwpGqX45kuYIbGY3X0JEOHJnbVU/view?usp=sharing

**Comment by obscuren** [comment]
Interesting. Thanks

**Comment by fjl** [comment]
Will look at the traces later.

**Comment by stephantual** [comment]
I have the same problem, upnp is on and working for for CppEth9.17, but not Geth9.19. 

If there's any log I can provide to help, I'd be happy to.

**Comment by deskenny** [comment]
Same here...

I0510 15:47:36.412071    5252 nat.go:94] network port 30303 could not be mapped: no UPnP or NAT-PMP router discovered

Dont remember having this problem a few months ago when I had Geth working on same ISP/router config.

**Comment by Ciemon** [comment]
Me too:

I0510 17:16:47.908137   23888 nat.go:94] network port 30303 could not be
mapped: no UPnP or NAT-PMP router discovered

Even though it's been working for weeks.

**Comment by deskenny** [comment]
mmmm... I got it going. Went into router and checked. Upnp was disabled (always must have been). Either the need is new, or I just didn't notice the error before. Anyway, works now for me

**Comment by obscuren** [comment]
> `network port 30303 could not be mapped: goupnp: SOAP request got HTTP 500 Internal Server Error`

Go to your router interface and disable / save and enable / save your UPnP settings on your router to restart the SOAP service on your router. 

> `network port 30303 could not be mapped: no UPnP or NAT-PMP router discovered`

**usually** means UPnP is disabled. Please make sure it's on.

**Comment by stephantual** [comment]
@obscuren 

```
upnpc : miniupnpc library test client. (c) 2006-2011 Thomas Bernard
Go to http://miniupnp.free.fr/ or http://miniupnp.tuxfamily.org/
for more information.
List of UPNP devices found on the network :
 desc: http://192.168.1.254:53393/rootDesc.xml
 st: urn:schemas-upnp-org:device:InternetGatewayDevice:1

Found valid IGD : http://192.168.1.254:53393/ctl/IPConn
Local LAN ip address : 192.168.1.169
Connection Type : PPPoE_Relay
Status : Connected, uptime=31001s, LastConnectionError : ERROR_NONE
  Time started : Sun May 10 16:32:46 2015
MaxBitRateDown : 4200000 bps (4.2 Mbps)   MaxBitRateUp 4200000 bps (4.2 Mbps)
ExternalIPAddress = 86.145.29.54
```

So armed with this, I removed the PPA (as I was unclear as to if it had been updated or not), and did

```
git clone https://github.com/ethereum/go-ethereum.git && cd go-ethereum && make geth 

geth

   Welcome to the
      FRONTIER

I0511 01:11:59.377327   12243 backend.go:236] Protocol Version: 60, Network Id: 0
I0511 01:11:59.377401   12243 backend.go:246] Blockchain DB Version: 2
I0511 01:11:59.378502   12243 chain_manager.go:235] Last block (#0) fd4af92a79c7fc2fd8bf0d342f2e832e1d4f485c85b9152d2039e03bc604fdca TD=0
I0511 01:11:59.384017   12243 cmd.go:107] Starting  Geth/v0.9.19-00280e62/linux/go1.4.2
I0511 01:11:59.384123   12243 server.go:209] Starting Server
I0511 01:12:01.467609   12243 nat.go:94] network port 30303 could not be mapped: no UPnP or NAT-PMP router discovered
I0511 01:12:01.474839   12243 udp.go:189] Listening, enode://4625c2043463abcbf3849dc1bb46e7cf4eef88156e5463979f18070dde7d5041bae4dd83caa54c48c236e78df6316603fb423bc8dbb911b46090dbac8d354175@86.145.29.54:30303
I0511 01:12:01.474901   12243 backend.go:450] Server started
```

````

**Comment by obscuren** [comment]
@stephantual thanks! @fjl PTAL

**Comment by stephantual** [comment]
FYI updated to latest devel commit and still no dice.

**Comment by fjl** [comment]
Will update this issue once I‎ have looked into it.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           ‎

**Comment by fjl** [comment]
Please note that UPNP is not required for connecting to the network.

**Comment by jpritikin** [comment]
> Please note that UPNP is not required for connecting to the network.

Curious, what is the benefit of a working UPNP?

**Comment by tgerring** [comment]
> Curious, what is the benefit of a working UPNP?

In theory, it enables your router to automatically allow inbound connections from other nodes to your local machine. This should result in better network connectivity and without it, you have to rely on making outbound connection and hoping those peers aren't behind a firewall or NAT that would normally block such traffic.

tldr;better connectivity resulting in healthier network

**Comment by fjl** [comment]
@jpritikin I have looked into the traces. Unfortunately, they only contain the multicast SSDP traffic. The bug seems to be in the part after the SSDP search, in the SOAP communication.

Would it be possible for you to do another capture including the unicast HTTP traffic?

**Comment by jpritikin** [comment]
@fjl Sure. Here is geth,

https://drive.google.com/file/d/0BwpGqX45kuYIdC1yc0loQW1uUGM/view?usp=sharing

Here is eth,

https://drive.google.com/file/d/0BwpGqX45kuYIaFFQRHVWaE04VTg/view?usp=sharing

Let me know if you need anything else.

**Comment by fjl** [comment]
@jpritikin please test again with the latest develop branch. the problem should be fixed now.

**Comment by jpritikin** [comment]
I can confirm that it appears to be working,

````
I0514 07:45:39.583452   17705 ssdp.go:63] ssdp: got unexpected search target result "urn:schemas-upnp-org:device:WANConnectionDevice:1"
I0514 07:45:39.583529   17705 ssdp.go:63] ssdp: got unexpected search target result "urn:schemas-upnp-org:device:WANConnectionDevice:1"
I0514 07:45:39.583570   17705 ssdp.go:63] ssdp: got unexpected search target result "urn:schemas-upnp-org:device:WANConnectionDevice:1"
I0514 07:45:39.621614   17705 udp.go:189] Listening, enode://b820c9f02cf880d162e24e42a55f868c3e69bd92324476eec675a52ca2bea43bdbc0be2def6be900b6d1a851c1724f11d48514dcb4cdadee53788fea47796729@73.171.34.52:30303
I0514 07:45:39.621740   17705 backend.go:456] Server started
I0514 07:45:39.621827   17705 server.go:349] Listening on [::]:30303
I0514 07:45:39.628702   17705 nat.go:95] mapped network port udp:30303 -> 30303 (ethereum discovery) using UPNP IGDv1-IP1
> I0514 07:45:39.649502   17705 nat.go:95] mapped network port tcp:30303 -> 30303 (ethereum p2p) using UPNP IGDv1-IP1
````

**Comment by obscuren** [comment]
:+1:

**Comment by stephantual** [comment]
can confirm this fixes it. 

```
24357 nat.go:95] mapped network port udp:30303 -> 30303 (ethereum discovery) using UPNP IGDv1-IP1
```

Thank you SO MUCH guys. This really helps

**Comment by stephantual** [comment]
Ah, spoke too soon, the client never gets any peers now.

**Comment by Tekarihoken** [comment]
I have observed exactly the same problem with geth 1.3.5, it indicate that it have mapped the udp port using UPNP but in reality the UPNP mapping is not visible in the router configuration and the synchro is drop down after few minutes.

**Comment by AugustoL** [comment]
Im Having the same problem, I cant connect to any peers. Im using homestead release, any fix?

**Comment by pipermerriam** [comment]
I'm having the same issue.

**Comment by AugustoL** [comment]
I find a way to fixed it, it seems that geth cant get the external ip, so I send it on the arguments like: --nat "extip:YOUR_PUBLIC_IP"

**Comment by leggomuhgreggo** [comment]
+1

**Comment by lssilva** [comment]
Same issue here:   1.4.9-stable-b7e3dfc5 any work around?

**Comment by sunujacob** [comment]
I get this message when running geth console.

> > E0313 08:49:44.752880 p2p/nat/nat.go:124] Network port udp:30303 could not be mapped: no UPnP or NAT-PMP router discovered


I wanted to know whether having this issue will cause problem for my node in syncing blockchain. I don't intend to do any mining on my node. I just want it use it as a wallet... create accounts, transfer ether and so on.

**Comment by REPTILEHAUS** [comment]
Also hitting this issue, --nat "extip:YOUR_PUBLIC_IP" suppresses the error message but nodes arent being discovered
